packages/nuxt/src/pages/runtime/router.options.ts

```diff
@@ -11,43 +11,70 @@ type ScrollPosition = Awaited<ReturnType<RouterScrollBehavior>>
+      window.history.scrollRestoration = 'auto'
+      return savedPosition
+    } else {
+      // For other navigations, ensure manual scrollRestoration
+      window.history.scrollRestoration = 'manual'
+    }
 
+    // Check if the route explicitly disables automatic scroll to top
     const routeAllowsScrollToTop = typeof to.meta.scrollToTop === 'function' ? to.meta.scrollToTop(to, from) : to.meta.scrollToTop
-
-    // Scroll to top if route is changed by default
-    if (!position && from && to && routeAllowsScrollToTop !== false && isChangingPage(to, from)) {
-      position = { left: 0, top: 0 }
+    if (routeAllowsScrollToTop === false) {
+      return false // Do not scroll to top if the route disallows it
     }
 
-    // Hash routes on the same page, no page hook is fired so resolve here
-    if (to.path === from.path) {
-      if (from.hash && !to.hash) {
-        return { left: 0, top: 0 }
+    const calculatePosition = (behavior: ScrollBehavior): ScrollPosition => {
+      // Handle saved position for backward/forward navigation
+      if (savedPosition) {
+        return savedPosition
       }
+
+      // Scroll to the element specified in the URL hash, if present
       if (to.hash) {
         return { el: to.hash, top: _getHashElementScrollMarginTop(to.hash), behavior }
       }
+
+      // Default scroll to the top left of the page
+      return { left: 0, top: 0, behavior }
+    }
+
+    // Handle same page navigation or first load
+    if (isFirstLoad || to.path === from.path) {
+      return calculatePosition(to.path === from.path || isFirstLoad ? scrollBehaviourType : 'instant')
+    }
+
+    // Check if the route change involves a page transition
+    const routeHasTransition = (route: RouteLocationNormalized): boolean => {
+      return !!(route.meta.pageTransition ?? defaultPageTransition) && !isFirstLoad
     }
+    const hasTransition = (routeHasTransition(from) && routeHasTransition(to))
 
-    // Wait for `page:transition:finish` or `page:finish` depending on if transitions are enabled or not
-    const hasTransition = (route: RouteLocationNormalized) => !!(route.meta.pageTransition ?? defaultPageTransition)
-    const hookToWait = (hasTransition(from) && hasTransition(to)) ? 'page:transition:finish' : 'page:finish'
+    // Wait for page transition to finish before applying the scroll behavior
+    const hookToWait = hasTransition ? 'page:transition:finish' : 'page:finish'
     return new Promise((resolve) => {
-      nuxtApp.hooks.hookOnce(hookToWait, async () => {
-        await nextTick()
-        if (to.hash) {
-          position = { el: to.hash, top: _getHashElementScrollMarginTop(to.hash), behavior }
-        }
-        resolve(position)
+      nuxtApp.hooks.hookOnce(hookToWait, () => {
+        return nextTick(() => {
+          // Without this setTimeout, the scroll behaviour is not applied, this is however working reliably
+          // todo: figure out how we can solve this without setTimeout of 0ms
```

Comment:
Maybe using `nextTick` twice here? 🤔

Reply:
I tried that, unfortunately it doesn't help. I even tried chaining multiple nextTicks() here and it didn't help. I think it has to do with what's happening when using setTimeout: https://medium.com/@sohnu/settimeout-with-time-0-what-does-it-really-mean-3b306880a0f6

Reply:
Should have added the quote I'm referring to:

> But what we have to see here is, all the 3 setTimeouts are taken away from the actual call executions and when there is nothing in the actual call stack, the setTimeouts started to execute with the time specified.
> 
> So whenever we want to take the call outside of the flow and run it once after the call stack is free, we can use setTimeout with time 0.

Reply:
I ran into an issue that had the same sort of resolution. I don't really understand what it is about `nextTick` that's not working in every case, but I opened a PR that might be relevant to this discussion: https://github.com/nuxt/nuxt/pull/25817